src/modm/driver/encoder/as5047.hpp

```diff
@@ -0,0 +1,152 @@
+// coding: utf-8
+// ----------------------------------------------------------------------------
+/*
+ * Copyright (c) 2024, Henrik Hose
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_AS5047_HPP
+#define MODM_AS5047_HPP
+
+#include <array>
+#include <modm/architecture/interface/register.hpp>
+#include <modm/architecture/interface/spi_device.hpp>
+#include <modm/processing/resumable.hpp>
+#include <modm/processing/timer.hpp>
+#include <numbers>
+
+namespace modm
+{
+
+constexpr uint16_t
+setMsbToEvenParity(const uint16_t num)
+{
+	uint16_t par = 0x7fff & num;
+	par ^= par >> 8;
+	par ^= par >> 4;
+	par ^= par >> 2;
+	par ^= par >> 1;
+	return ((par & 1) << 15) | (0x7fff & num);
+}
```

Comment:
This is the modm namespace which is quite generic. Would be good to move to `struct as5047` or put into the relevant `:math` module.

Reply:
I'm not sure, how to move it into the struct?

<details>
<summary>
If I declare `setMsbToEvenParity` as `static constexpr` in the `as5047` struct,
I get an error saying `error: 'static constexpr uint16_t modm::as5047::setMsbToEvenParity(uint16_t)' used before its definition`
</summary>

```
struct as5047
{
	enum class Errorfl : uint16_t
	{
		Parerr = Bit2,
		Invcomm = Bit1,
		Frerr = Bit0,
	};
	MODM_FLAGS16(Errorfl)

	enum class Prog : uint16_t
	{
		Progver = Bit6,
		Progotp = Bit3,
		Otpref = Bit2,
		Progen = Bit0,
	};
	MODM_FLAGS16(Prog)

	enum class Diaagc : uint16_t
	{
		Magl = Bit11,
		Magh = Bit10,
		Cof = Bit9,
		Lf = Bit8,
	};
	MODM_FLAGS16(Diaagc)

	static constexpr uint16_t
	setMsbToEvenParity(const uint16_t num)
	{
		uint16_t par = 0x7fff & num;
		par ^= par >> 8;
		par ^= par >> 4;
		par ^= par >> 2;
		par ^= par >> 1;
		return ((par & 1) << 15) | (0x7fff & num);
	}

	enum class Register : uint16_t
	{
		ReadNop = setMsbToEvenParity(((1 << 14) | 0x0000)),
		ReadErrfl = setMsbToEvenParity(((1 << 14) | 0x0001)),
		ReadProg = setMsbToEvenParity(((1 << 14) | 0x0003)),
		ReadDiaagc = setMsbToEvenParity(((1 << 14) | 0x3FFC)),
		ReadMag = setMsbToEvenParity(((1 << 14) | 0x3FFD)),
		ReadAngleunc = setMsbToEvenParity(((1 << 14) | 0x3FFE)),
		ReadAnglecom = setMsbToEvenParity(((1 << 14) | 0x3FFF)),

	};

	struct modm_packed Data
	{
		/// @return
		constexpr float
		getAngleRad() const
		{
			const uint16_t angle = data & 0x3fff;
			return static_cast<float>(angle) / 16383.f * 2.f * std::numbers::pi_v<float>;
		}

		/// @return
		constexpr float
		getAngleDeg() const
		{
			const uint16_t angle = data & 0x3fff;
			return static_cast<float>(angle) / 16383.f * 360.f;
		}

		/// @return
		constexpr uint16_t
		getAngleRaw() const
		{
			const uint16_t angle = data & 0x3fff;
			return angle;
		}

		uint16_t data;
	};
};  // struct as5047
```
</details>

I think this might be related to [this](https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#1626): 
> The Standard should make clear that a constexpr member function cannot be used in a constant expression until its class is complete.

Do you have any suggestions?
Should I put everything into an `as5047` namespace?

I don't think moving the function as is to the `math:` is appropriate, since the function specifically calculates and sets bit 15 of a `uint16_t`, if moved to `math:`, a more generic version might be more appropriate. Also, other modm modules like `etlcpp` have their own functions to calculate parity bits...

Reply:
Ugh, C++ amiright!1!!

You can put it into a `modm::details` namespace then and use `/// @cond` and `/// @endcond` to block doxypress from indexing it ([Examples](https://github.com/search?q=repo%3Amodm-io%2Fmodm+%22namespace+detail%22+path%3A%2F%5Esrc%5C%2Fmodm%5C%2F%2F&type=code)). Prolly prefix the function with `as5047_` to avoid collisions if you wanna be super careful.

Reply:
> Ugh, C++ amiright!1!!

:cat2: No we are super careful :cat: